### ISSUE TITLE
alg

### ISSUE THREAD

**Issue by 9535299035** [9535299035/eeg-monitor#3]
import numpy as np

from scipy.signal import butter, filtfilt, iirnotch, welch
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler

class EEG_BCI:

    def __init__(self, sampling_rate=250):

        self.fs = sampling_rate

        self.scaler = StandardScaler()

        self.classifier = LinearDiscriminantAnalysis()

        self.trained = False

    # ------------------------------------------------
    # FILTER
    # ------------------------------------------------

    def bandpass(self, eeg):

        nyquist = self.fs / 2

        b, a = butter(
            4,
            [1 / nyquist, 40 / nyquist],
            btype="band"
        )

        return filtfilt(
            b,
            a,
            eeg
        )

    # ------------------------------------------------
    # REMOVE 50 Hz NOISE
    # ------------------------------------------------

    def notch(self, eeg):

        b, a = iirnotch(
            50,
            30,
            self.fs
        )

        return filtfilt(
            b,
            a,
            eeg
        )

    # ------------------------------------------------
    # PREPROCESS EEG
    # ------------------------------------------------

    def preprocess(self, eeg):

        eeg = np.asarray(
            eeg,
            dtype=float
        )

        # Remove DC component
        eeg = eeg - np.mean(
            eeg,
            axis=-1,
            keepdims=True
        )

        # Bandpass
        eeg = self.bandpass(
            eeg
        )

        # Notch
        eeg = self.notch(
            eeg
        )

        return eeg

    # ------------------------------------------------
    # FREQUENCY FEATURES
    # ------------------------------------------------

    def frequency_features(
        self,
        eeg
    ):

        frequencies, power = welch(
            eeg,
            fs=self.fs,
            axis=-1,
            nperseg=min(
                eeg.shape[-1],
                256
            )
        )

        bands = {

            "delta": (1, 4),

            "theta": (4, 8),

            "alpha": (8, 13),

            "beta": (13, 30),

            "gamma": (30, 40)
        }

        features = []

        for low, high in bands.values():

            mask = (
                (frequencies >= low) &
                (frequencies < high)
            )

            band_power = np.trapezoid(
                power[..., mask],
                frequencies[mask],
                axis=-1
            )

            features.append(
                band_power
            )

        return np.concatenate(
            features,
            axis=-1
        )

    # ------------------------------------------------
    # FEATURE EXTRACTION
    # ------------------------------------------------

    def extract_features(self, eeg):

        cleaned = self.preprocess(
            eeg
        )

        features = self.frequency_features(
            cleaned
        )

        return features

    # ------------------------------------------------
    # TRAIN BCI
    # ------------------------------------------------

    def train(
        self,
        eeg_training_data,
        labels
    ):

        feature_data = []

        for eeg in eeg_training_data:

            features = self.extract_features(
                eeg
            )

            feature_data.append(
                features
            )

        X = np.asarray(
            feature_data
        )

        X = self.scaler.fit_transform(
            X
        )

        self.classifier.fit(
            X,
            labels
        )

        self.trained = True

        print(
            "BCI training complete."
        )

    # ------------------------------------------------
    # PREDICT
    # ------------------------------------------------

    def predict(self, eeg):

        if not self.trained:

            raise RuntimeError(
                "BCI has not been trained."
            )

        features = self.extract_features(
            eeg
        )

        features = self.scaler.transform(
            features.reshape(1, -1)
        )

        prediction = self.classifier.predict(
            features
        )[0]

        probability = self.classifier.predict_proba(
            features
        )[0]

        confidence = np.max(
            probability
        )

        return prediction, confidence

    # ------------------------------------------------
    # COMMAND DECODER
    # ------------------------------------------------

    def command(self, prediction, confidence):

        if confidence < 0.70:

            return "NO COMMAND"

        commands = {

            0: "LEFT",

            1: "RIGHT",

            2: "FORWARD",

            3: "BACKWARD",

            4: "STOP"
        }

        return commands.get(
            prediction,
            "UNKNOWN"
        ) create an (*kemmelrgmail.com )
open 50 hrs
         70haz 
         60hrz
        import numpy as np
from scipy.signal import butter, filtfilt, welch

class GammaWaveDetector:

    def __init__(self, sampling_rate=500):

        self.fs = sampling_rate

        # Gamma range
        self.low_gamma = 30
        self.high_gamma = 100

    def bandpass_gamma(self, eeg):

        nyquist = self.fs / 2

        low = self.low_gamma / nyquist
        high = self.high_gamma / nyquist

        b, a = butter(
            4,
            [low, high],
            btype="band"
        )

        return filtfilt(
            b,
            a,
            eeg
        )

    def gamma_power(self, eeg):

        frequencies, power = welch(
            eeg,
            fs=self.fs,
            nperseg=min(
                len(eeg),
                self.fs * 2
            )
        )

        mask = (
            (frequencies >= 30) &
            (frequencies <= 100)
        )

        gamma_power = np.trapezoid(
            power[mask],
            frequencies[mask]
        )

        return gamma_power

    def detect(self, eeg):

        gamma_signal = self.bandpass_gamma(
            eeg
        )

        power = self.gamma_power(
            eeg
        )

        return {
            "gamma_signal": gamma_signal,
            "gamma_power": power
        }

# --------------------------------------------------
# Example
# --------------------------------------------------

FS = 500

detector = GammaWaveDetector(
    sampling_rate=FS
)

# Example EEG signal
time = np.arange(
    0,
    2,
    1 / FS
)

eeg = (
    0.5 * np.random.randn(len(time))
    +
    0.2 * np.sin(
        2 * np.pi * 40 * time
    )
)

result = detector.detect(
    eeg
)

print(
    "Gamma Power:",
    result["gamma_power"]
)open with true 
with gmail 
import numpy as np
import matplotlib.pyplot as plt

def calculate_frequency(eeg_signal, sampling_rate):

    # Number of samples
    N = len(eeg_signal)

    # Remove DC offset
    eeg_signal = eeg_signal - np.mean(eeg_signal)

    # FFT
    fft_result = np.fft.rfft(eeg_signal)

    # Frequency axis
    frequencies = np.fft.rfftfreq(
        N,
        1 / sampling_rate
    )

    # Magnitude
    magnitude = np.abs(fft_result) / N

    return frequencies, magnitude

# ------------------------------------------------
# CREATE TEST EEG SIGNAL
# ------------------------------------------------

sampling_rate = 500

duration = 5

time = np.arange(
    0,
    duration,
    1 / sampling_rate
)

# Example EEG containing several frequencies
eeg = (
    0.5 * np.sin(2 * np.pi * 10 * time) +
    0.3 * np.sin(2 * np.pi * 40 * time) +
    0.2 * np.random.randn(len(time))
)

# ------------------------------------------------
# CALCULATE FREQUENCY
# ------------------------------------------------

frequency, magnitude = calculate_frequency(
    eeg,
    sampling_rate
)

# ------------------------------------------------
# DISPLAY
# ------------------------------------------------

plt.figure(figsize=(12, 5))

plt.plot(
    frequency,
    magnitude
)
****
plt.xlim(
    0,
    100
)

plt.xlabel(
    "Frequency (Hz)"
)

plt.ylabel(
    "Magnitude"
)

plt.title(
    "EEG Frequency Spectrum"
)

plt.grid()

plt.show()
